### PR TITLE
Support StringView for binary operators

### PR DESCRIPTION
**Pull request by tlm365**
## Which issue does this PR close?
Part of #12180.

## Rationale for this change
Support StringView for binary operators like `~`, `!~`, etc.

## What changes are included in this PR?
If Utf8View is in any side of binary operators, we COERCE to Utf8. This behaviour will remain until native StringView support is implemented.

## Are these changes tested?
Yes.

## Are there any user-facing changes?
No.

### CODE REVIEW

**Comment by dharanad** [comment]
I plan to review this later today.

File: datafusion/sqllogictest/test_files/string_view.slt [hunk 1]
```diff
@@ -1225,6 +1221,48 @@ XiangpengXiangpeng XiangpengXiangpeng XiangpengXiangpeng
 RaphaelRaphael RaphaelRaphael RaphaelRaphael
 NULL NULL NULL
 
+## Ensure no casts for binary operators
+## TODO: https://github.com/apache/datafusion/issues/12180
+# `~` operator (regex match)
+query TT
+EXPLAIN SELECT
```
Comment (tshauck) [inline datafusion/sqllogictest/test_files/string_view.slt:1228]: Just parroting some feedback I got on a similar PR: might be nice to include the queries w/o `EXPLAIN` so as to verify correctness not just casts (or in the future no casts 😄).

Comment (tlm365) [reply datafusion/sqllogictest/test_files/string_view.slt:1228]: @tshauck Thanks for reviewing. Makes sense to me, I will update it soon.

**Review by alamb — APPROVED** [review]
Thank you @tlm365  and @tshauck  -- this looks good to me. 

Supporting the operators at all is a good first step and then we can avoid the cast to utf8 as a follow on

**Pull request merged**